src/python/pants/engine/rules.py

```diff
@@ -375,57 +371,35 @@ def __new__(cls,
       )
 
   def __str__(self):
-    return '({}, {!r}, {})'.format(type_or_constraint_repr(self.output_constraint),
+    return '({}, {!r}, {})'.format(self.output_type.__name__,
                                    self.input_selectors,
                                    self.func.__name__)
 
 
-class SingletonRule(datatype(['output_constraint', 'value']), Rule):
+class SingletonRule(datatype([('output_type', _type_field), 'value']), Rule):
   """A default rule for a product, which is thus a singleton for that product."""
 
   @classmethod
   def from_instance(cls, obj):
     return cls(type(obj), obj)
 
-  def __new__(cls, output_type, value):
-    # Validate result type.
-    if isinstance(output_type, Exactly):
-      constraint = output_type
-    elif isinstance(output_type, type):
-      constraint = Exactly(output_type)
-    else:
-      raise TypeError("Expected an output_type for rule; got: {}".format(output_type))
-
-    # Create.
-    return super(SingletonRule, cls).__new__(cls, constraint, value)
 
-  @property
-  def dependency_optionables(self):
-    return tuple()
-
-  def __repr__(self):
-    return '{}({}, {})'.format(type(self).__name__, type_or_constraint_repr(self.output_constraint), self.value)
-
-
-class RootRule(datatype(['output_constraint']), Rule):
+class RootRule(datatype([('output_type', _type_field)]), Rule):
   """Represents a root input to an execution of a rule graph.
 
   Roots act roughly like parameters, in that in some cases the only source of a
   particular type might be when a value is provided as a root subject at the beginning
   of an execution.
   """
 
-  @property
-  def dependency_optionables(self):
-    return tuple()
-
 
 class RuleIndex(datatype(['rules', 'roots'])):
   """Holds a normalized index of Rules used to instantiate Nodes."""
 
   @classmethod
   def create(cls, rule_entries):
     """Creates a RuleIndex with tasks indexed by their output type."""
+    # TODO: make a defaultdict-like wrapper for OrderedDict (and other types?)!
```

Comment:
Sortof off-topic I think.

Reply:
Ended up making #7311 and moving it down to the part where we index into an OrderedDict since if we do more stuff like in #7116 we probably do want to do something like that.